tests/D2L.CodeStyle.Analyzers.Test/Specs/ImmutabilityAnalyzer.cs

```diff
@@ -315,4 +326,86 @@ sealed class ClassWithExceptedImmutableMemberSucceeds {
+			internal readonly GenericClassWithImmutableState<MutableClass> m_ConcreteClassWithMutableGenericType;
+		}
+
+		[Objects.Immutable]
+		class ConcreteClassWithImmutableGenericType {
+			internal readonly GenericClassWithImmutableState<ImmutableClass> m_ConcreteClassWithImmutableGenericType;
+		}
+	}
+
+	class GenericTypeParameterTests {
+		// Ensures generic type parameters from an interface are examined
+
+		[Objects.Immutable]
+		interface ImmutableGenericInterface<[Objects.Immutable] T> {
+		}
+
+		[Objects.Immutable]
+		sealed class GenericClassWithImmutableInterface<T> : ImmutableGenericInterface<T> {
+			internal readonly T m_GenericClassWithImmutableInterface;
+		}
+
+		[Objects.Immutable]
+		class ConcreteClassWithImmutableInterfaceImmutableType {
+			internal readonly ImmutableGenericInterface<ImmutableClass> m_ConcreteClassWithImmutableInterfaceImmutableType;
+		}
+
+		[Objects.Immutable]
+		class /* ImmutableClassIsnt('m_ConcreteClassWithImmutableInterfaceMutableType''s type ('SpecTests.MutableClass') is a type parameter that must be marked with `[Objects.Immutable]`) */ ConcreteClassWithImmutableInterfaceMutableType /**/ {
+			internal readonly ImmutableGenericInterface<MutableClass> m_ConcreteClassWithImmutableInterfaceMutableType;
+		}
+	}
+
+	class MultipleGenericParameterTests {
+		// Tests when there is a mixed set of mutable and immutable type parameters
+
+		[Objects.Immutable]
+		sealed class MultiGenericClassWithOneImmutable<[Objects.Immutable] T, S> {
+			internal readonly T m_MultiGenericClassWithOneImmutable;
+		}
+
+		[Objects.Immutable]
+		sealed class ConcreteHoldingMixedModeGeneric {
+			internal readonly MultiGenericClassWithOneImmutable<ImmutableClass, MutableClass> m_ConcreteHoldingMixedModeGeneric;
+		}
+
+		[Objects.Immutable]
+		sealed class /* ImmutableClassIsnt('m_ConcreteHoldingMixedModeGenericWrongOrder.m_MultiGenericClassWithOneImmutable.m_MutableClass' is not read-only) */ ConcreteHoldingMixedModeGenericWrongOrder /**/ {
+			internal readonly MultiGenericClassWithOneImmutable<MutableClass, ImmutableClass> m_ConcreteHoldingMixedModeGenericWrongOrder;
+		}
+	}
+
+	class MultipleGenericInterfaceParameterTests {
+		[Objects.Immutable]
+		interface MultiGeneric<[Objects.Immutable] S, T> {
+		}
+
+		sealed class MultiGenericFromInterface<S, T> : MultiGeneric<S, T> {
+			internal readonly S m_MultiGenericFromInterface;
+		}
+
+		[Objects.Immutable]
+		sealed class ConcreteUsingMultiGenericFromInterface {
+			internal readonly MultiGenericFromInterface<ImmutableClass, NonReadOnlyClass> m_ConcreteUsingMultiGenericFromInterface;
+		}
+
+		[Objects.Immutable]
+		sealed class /* ImmutableClassIsnt('m_ConcreteUsingMultiGenericFromInterfaceWrongOrder.m_MultiGenericFromInterface.m_MutableClass' is not read-only) */ ConcreteUsingMultiGenericFromInterfaceWrongOrder /**/ {
+			internal readonly MultiGenericFromInterface<MutableClass, ImmutableClass> m_ConcreteUsingMultiGenericFromInterfaceWrongOrder;
+		}
+	}

```

Comment:
Other test ideas:
* What happens if you add `Objects.Immutable` to type parameter, when it is superfluous?  Ie, nothing actually holds on to that type.  Does it do nothing (since the implementations are always immutable), or complain if a mutable type is used? (Someone will make a refactoring mistake someday)
* Could we add a test where more than one type parameter is marked `Immutable`?  Probably works as intended, but I'd like the safety net.